cluster-autoscaler/utils/gpu/gpu_test.go

```diff
@@ -31,54 +34,191 @@ const (
+			desc: "missing gpu",
+			config: &cloudprovider.GpuConfig{
+				Label:        "label/gpu-type",
+				Type:         "gpu-type-value",
+				ResourceName: util.ResourceNvidiaGPU,
+			},
+			gpus: map[string]struct{}{
+				"gpu-type-value": {},
+			},
+			node:         newNode("label/gpu-type", "gpu-type-value", nil),
+			templateNode: newNode("label/gpu-type", "gpu-type-value", resource.NewQuantity(10, resource.DecimalSI)),
+			wantResource: util.ResourceNvidiaGPU,
+			wantLabel:    "missing-gpu",
 		},
-		Status: apiv1.NodeStatus{
-			Capacity:    apiv1.ResourceList{},
-			Allocatable: apiv1.ResourceList{},
+		{
+			desc: "no capacity",
+			config: &cloudprovider.GpuConfig{
+				Label:        "label/gpu-type",
+				Type:         "gpu-type-value",
+				ResourceName: util.ResourceNvidiaGPU,
+			},
+			gpus: map[string]struct{}{
+				"gpu-type-value": {},
+			},
+			node:         newNode("label/gpu-type", "gpu-type-value", nil),
+			templateNode: newNode("label/gpu-type", "gpu-type-value", nil),
+			wantResource: util.ResourceNvidiaGPU,
+			wantLabel:    "unexpected-label",
+		},
+		{
+			desc: "no capacity and node group",
+			config: &cloudprovider.GpuConfig{
+				Label:        "label/gpu-type",
+				Type:         "gpu-type-value",
+				ResourceName: util.ResourceNvidiaGPU,
+			},
+			gpus: map[string]struct{}{
+				"gpu-type-value": {},
+			},
+			node:         newNode("label/gpu-type", "gpu-type-value", nil),
+			wantResource: util.ResourceNvidiaGPU,
+			wantLabel:    "unexpected-label",
 		},
 	}
-	assert.False(t, NodeHasGpu(GPULabel, nodeNoGpu))
-}
+	for _, tc := range tests {
+		func(tc testCase) {
+			t.Run(tc.desc, func(t *testing.T) {
+				var nodeGroup cloudprovider.NodeGroup
+				if tc.templateNode != nil {
+					var nodeInfo = schedulerframework.NewNodeInfo()
+					nodeInfo.SetNode(tc.templateNode)
 
-func TestPodRequestsGpu(t *testing.T) {
-	podNoGpu := test.BuildTestPod("podNoGpu", 0, 1000)
-	podWithGpu := test.BuildTestPod("pod1AnyGpu", 0, 1000)
-	podWithGpu.Spec.Containers[0].Resources.Requests[ResourceNvidiaGPU] = *resource.NewQuantity(1, resource.DecimalSI)
+					provider := testprovider.NewTestAutoprovisioningCloudProvider(
+						func(_ string, _ int) error { return nil },
+						func(_ string, _ string) error { return nil },
+						func(_ string) error { return nil },
+						func(_ string) error { return nil },
+						[]string{},
+						map[string]*schedulerframework.NodeInfo{
+							"my-node-group": nodeInfo,
+						})
 
-	assert.False(t, PodRequestsGpu(podNoGpu))
-	assert.True(t, PodRequestsGpu(podWithGpu))
+					provider.AddNodeGroup("my-node-group", 0, 100, 1)
+					nodeGroup = provider.GetNodeGroup("my-node-group")
+				}
+
+				gotResource, gotLabel := GetGpuInfoForMetrics(tc.config, tc.gpus, tc.node, nodeGroup)
+
+				assert.Equal(t, tc.wantResource, gotResource)
+				assert.Equal(t, tc.wantLabel, gotLabel)
+			})
+		}(tc)
+	}
+}
+
+func newNode(gpuLabel, gpuType string, gpuCapacity *resource.Quantity) *apiv1.Node {
```

Comment:
The current `GetGpuInfoForMetrics` implementation uses only the node capacity, not the labels. The gpu label/value is extracted directly from the `GpuConfig`. I think it would be better to change this method to `newNode(gpuCapacity *resource.Quantity) *apiv1.Node` and not set the labels at all. This way we also document that labels/values are taken from `GpuConfig`

Reply:
Done